### PR TITLE
Add: Metadata report base

### PR DESCRIPTION
**Pull request by hxmhlt**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hxmhlt** [comment]
#418 #444

File: config/service_config.go [hunk 1]
```diff
@@ -245,3 +275,68 @@ func (c *ServiceConfig) getUrlMap() url.Values {
...
+			url := exporter.GetInvoker().GetUrl()
+			urls = append(urls, &url)
+		}
+		return urls
+	}
+	return nil
+}
+
+// MockInitProviderWithSingleRegistry will init a mocked providerConfig
+func MockInitProviderWithSingleRegistry() {
```
Comment (flycash) [inline config/service_config.go:293]: Could you move the mock data into test file or some mock data file?

File: metadata/service/exporter/configurable/exporter.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	exporter.lock.RLock()
+	defer exporter.lock.RUnlock()
+	return exporter.serviceConfig != nil && exporter.serviceConfig.IsExport()
+}
+
+// generateMetadataProtocol will return a default ProtocolConfig
+func generateMetadataProtocol() *config.ProtocolConfig {
+	return &config.ProtocolConfig{
+		Name: constant.DEFAULT_PROTOCOL,
+		Port: "20000",
```
Comment (flycash) [inline metadata/service/exporter/configurable/exporter.go:98]: Is necessary to make this Port configurable??

File: metadata/service/service.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/emirpasic/gods/sets"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/config"
+)
+
+// Metadataservice is used to define meta data related behaviors
+type MetadataService interface {
```
Comment (flycash) [inline metadata/service/service.go:30]: Add comments for each method.

File: metadata/namemapping/service_name_mapping.go [hunk 4]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package metadata
+package namemapping
```
Comment (zouyx) [inline metadata/namemapping/service_name_mapping.go:18]: Could you use one word for this package name

File: metadata/identifier/base_metadata_identifier.go [hunk 5]
```diff
@@ -47,25 +49,25 @@ func joinParams(joinChar string, params []string) string {
...
 	return constant.DEFAULT_PATH_TAG +
 		withPathSeparator(path) +
-		withPathSeparator(mdi.version) +
-		withPathSeparator(mdi.group) +
-		withPathSeparator(mdi.side) +
-		joinParams(constant.PATH_SEPARATOR, params)
+		withPathSeparator(mdi.Version) +
+		withPathSeparator(mdi.Group) +
+		withPathSeparator(mdi.Side) +
+		joinParams("/", params)
```
Comment (zouyx) [inline metadata/identifier/base_metadata_identifier.go:70]: Would you consider use constant for ```/``` ?

File: metadata/identifier/base_metadata_identifier.go [hunk 6]
```diff
@@ -86,7 +88,7 @@ func serviceToPath(serviceInterface string) string {
 //withPathSeparator...
 func withPathSeparator(path string) string {
 	if len(path) != 0 {
-		path = constant.PATH_SEPARATOR + path
+		path = "/" + path
```
Comment (zouyx) [inline metadata/identifier/base_metadata_identifier.go:91]: as above

File: metadata/definition/definition.go [hunk 7]
```diff
@@ -39,3 +50,39 @@ type TypeDefinition struct {
...
+
+	return sd
+}
+
+// ServiceDescriperBuild: build the service key, format is `group/serviceName:version` which be same as URL's service key
+func ServiceDescriperBuild(serviceName string, group string, version string) string {
+	buf := &bytes.Buffer{}
+	if group != "" {
+		buf.WriteString(group)
+		buf.WriteString("/")
```
Comment (zouyx) [inline metadata/definition/definition.go:80]: Would you consider use constant for ```/``` ?

File: metadata/definition/definition.go [hunk 8]
```diff
@@ -39,3 +50,39 @@ type TypeDefinition struct {
...
+	}
+	buf.WriteString(serviceName)
+	if version != "" && version != "0.0.0" {
+		buf.WriteString(":")
```
Comment (zouyx) [inline metadata/definition/definition.go:84]: Would you consider use constant for ```:``` ?

File: config/service_config.go [hunk 9]
```diff
@@ -160,22 +172,40 @@ func (c *ServiceConfig) Export() error {
...
 
+// Unexport will call unexport of all exporters service config exported
+func (c *ServiceConfig) Unexport() {
+	if !c.exported.Load() {
+		return
+	}
+	if c.unexported.Load() {
+		return
+	}
+	for _, exporter := range c.exporters {
```
Comment (zouyx) [inline config/service_config.go:201]:
```suggestion
        e:=c.exporters
	for _, exporter := range e {
```

how about this?

File: config/service_config.go (hunk 1 above)
Comment (zouyx) [reply config/service_config.go:293]: agree

File: metadata/service/exporter/configurable/exporter.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+		exporter.serviceConfig.Group = config.GetApplicationConfig().Name
+		exporter.serviceConfig.Version = exporter.metadataService.Version()
+		exporter.serviceConfig.Implement(exporter.metadataService)
+		err := exporter.serviceConfig.Export()
+		logger.Infof("The MetadataService exports urls : %v ", exporter.serviceConfig.GetExportedUrls())
+		return err
+	} else {
+		logger.Warnf("The MetadataService has been exported : %v ", exporter.serviceConfig.GetExportedUrls())
+		return nil
+	}
```
Comment (zouyx) [inline metadata/service/exporter/configurable/exporter.go:68]:
```suggestion
	} 
		logger.Warnf("The MetadataService has been exported : %v ", exporter.serviceConfig.GetExportedUrls())
		return nil

```

File: metadata/service/exporter/exporter.go [hunk 11]
```diff
@@ -15,15 +15,15 @@
...
 import (
 	"github.com/apache/dubbo-go/common"
 )
 
-type MetadataExporter interface {
-	Export() MetadataExporter
-	Unexport() MetadataExporter
+type MetadataServiceExporter interface {
+	Export() error
+	Unexport()
```
Comment (zouyx) [inline metadata/service/exporter/exporter.go:26]: no error?

File: metadata/service/service.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+// ServiceName can get the service's name in meta service , which is application name
+func (mts *BaseMetadataService) ServiceName() (string, error) {
+	return config.GetApplicationConfig().Name, nil
+}
+
+// RefreshMetadata is used for event listener's calling, to refresh metadata
+//func (mts *BaseMetadataService) RefreshMetadata(exportedRevision string, subscribedRevision string) bool {
+//	return true
+//}
```
Comment (zouyx) [inline metadata/service/service.go:59]: delete it if it is useless

File: metadata/service/exporter/configurable/exporter.go (hunk 2 above)
Comment (hxmhlt) [reply metadata/service/exporter/configurable/exporter.go:98]: Only after the random port feature is supported

File: metadata/namemapping/service_name_mapping.go (hunk 4 above)
Comment (hxmhlt) [reply metadata/namemapping/service_name_mapping.go:18]: maybe namemapping is one word?

File: config/service_config.go (hunk 9 above)
Comment (hxmhlt) [reply config/service_config.go:201]: why?

File: config/config_loader_test.go [hunk 13]
```diff
@@ -233,3 +234,55 @@ func TestConfigLoaderWithConfigCenterSingleRegistry(t *testing.T) {
...
 }
+
+// mockInitProviderWithSingleRegistry will init a mocked providerConfig
+func mockInitProviderWithSingleRegistry() {
+	providerConfig = &ProviderConfig{
+		ApplicationConfig: &ApplicationConfig{
+			Organization: "dubbo_org",
+			Name:         "dubbo",
+			Module:       "module",
+			Version:      "2.6.0",
```
Comment (AlexStocks) [inline config/config_loader_test.go:245]: maybe u can change the version to 2.7.4.

File: metadata/identifier/service_metadata_identifier.go [hunk 14]
```diff
@@ -21,18 +21,19 @@ import (
...
+// ServiceMetadataIdentifier is inherit baseMetaIdentifier with service params: Revision and Protocol
 type ServiceMetadataIdentifier struct {
-	revision string
-	protocol string
+	Revision string
+	Protocol string
 	BaseMetadataIdentifier
 }
 
 // getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go:31]: pls add the comment.

File: metadata/identifier/service_metadata_identifier.go [hunk 15]
```diff
@@ -21,18 +21,19 @@ import (
...
-func (mdi *ServiceMetadataIdentifier) getIdentifierKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.protocol + constant.KEY_REVISON_PREFIX + mdi.revision)
+func (mdi *ServiceMetadataIdentifier) GetIdentifierKey() string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.Protocol, constant.KEY_REVISON_PREFIX+mdi.Revision)
 }
 
 // getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go:36]: pls add the comment.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 16]
```diff
@@ -17,17 +17,18 @@
...
 package identifier
 
+// SubscriberMetadataIdentifier is inherit baseMetaIdentifier with service params: Revision
 type SubscriberMetadataIdentifier struct {
-	revision string
+	Revision string
 	BaseMetadataIdentifier
 }
 
 // getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go:26]: pls add the comment.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 17]
```diff
@@ -17,17 +17,18 @@
...
-func (mdi *SubscriberMetadataIdentifier) getIdentifierKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.revision)
+func (mdi *SubscriberMetadataIdentifier) GetIdentifierKey() string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.Revision)
 }
 
 // getIdentifierKey...
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go:31]: pls add the comment.

File: metadata/report/factory/report_factory.go [hunk 18]
```diff
@@ -15,16 +15,20 @@
...
-package metadata
+package factory
 
 import (
 	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/metadata/report"
 )
 
 var (
-	MetadataReportInstance MetadataReport
```
Comment (AlexStocks) [inline metadata/report/factory/report_factory.go:25]: comment, pls.

File: metadata/report/factory/report_factory.go [hunk 19]
```diff
@@ -15,16 +15,20 @@
...
+	MetadataReportInstance report.MetadataReport
 )
 
 type MetadataReportFactory interface {
```
Comment (AlexStocks) [inline metadata/report/factory/report_factory.go:29]: comment, pls.

File: metadata/report/factory/report_factory.go [hunk 20]
```diff
@@ -15,16 +15,20 @@
...
-	CreateMetadataReport(*common.URL) MetadataReport
+	CreateMetadataReport(*common.URL) report.MetadataReport
```
Comment (AlexStocks) [inline metadata/report/factory/report_factory.go:30]: comment, pls.

File: metadata/service/exporter/exporter.go [hunk 21]
```diff
@@ -15,15 +15,15 @@
...
+package exporter
 
 import (
 	"github.com/apache/dubbo-go/common"
 )
 
-type MetadataExporter interface {
-	Export() MetadataExporter
-	Unexport() MetadataExporter
+type MetadataServiceExporter interface {
```
Comment (AlexStocks) [inline metadata/service/exporter/exporter.go:24]: comment, pls.

File: metadata/service/inmemory/service.go [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+	exportedServiceURLs   *sync.Map
+	subscribedServiceURLs *sync.Map
+	serviceDefinitions    *sync.Map
+	lock                  *sync.RWMutex
+}
+
+// NewMetadataService: initiate a metadata service
+func NewMetadataService() *MetadataService {
+	return &MetadataService{
+		exportedServiceURLs:   new(sync.Map),
```
Comment (AlexStocks) [inline metadata/service/inmemory/service.go:50]: pls do not use new. the `&sync.Map` should be ok.

File: metadata/service/inmemory/service.go [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+		subscribedServiceURLs: new(sync.Map),
+		serviceDefinitions:    new(sync.Map),
+		lock:                  new(sync.RWMutex),
```
Comment (AlexStocks) [inline metadata/service/inmemory/service.go:53]: &sync.RWMutex

File: metadata/service/exporter/exporter.go (hunk 11 above)
Comment (hxmhlt) [reply metadata/service/exporter/exporter.go:26]: yeah

File: metadata/report/report.go [hunk 24]
```diff
@@ -15,14 +15,15 @@
...
 import (
 	"github.com/apache/dubbo-go/common"
 	"github.com/apache/dubbo-go/metadata/definition"
 	"github.com/apache/dubbo-go/metadata/identifier"
 )
 
+// MetadataReport is an interface of remote metadata report
 type MetadataReport interface {
 	StoreProviderMetadata(*identifier.MetadataIdentifier, *definition.ServiceDefinition)
 	StoreConsumeretadata(*identifier.MetadataIdentifier, map[string]string)
```
Comment (gaoxinge) [inline metadata/report/report.go:29]: Fix typo. It should be `StoreConsumerMetadata`.

File: common/rpc_service.go [hunk 25]
```diff
@@ -133,6 +133,11 @@ func (s *Service) Method() map[string]*MethodType {
 	return s.methods
 }
 
+// Name will return service name
+func (s *Service) Name() string {
+	return s.name
+}
+
 // RcvrType ...
```
Comment (AlexStocks) [inline common/rpc_service.go:141]: pls add comment for this file. thx.

File: config/service_config.go [hunk 26]
```diff
@@ -71,11 +71,14 @@ type ServiceConfig struct {
...
 	Tag                         string            `yaml:"tag" json:"tag,omitempty" property:"tag"`
 
+	Protocols     map[string]*ProtocolConfig
 	unexported    *atomic.Bool
 	exported      *atomic.Bool
 	rpcService    common.RPCService
 	cacheProtocol protocol.Protocol
 	cacheMutex    sync.Mutex
+	exporters     []protocol.Exporter
+	exportersLock sync.Mutex
```
Comment (AlexStocks) [inline config/service_config.go:81]: pls move the lock above its target member

File: metadata/identifier/metadata_identifier.go [hunk 27]
```diff
@@ -17,17 +17,18 @@
...
 type MetadataIdentifier struct {
-	application string
+	Application string
 	BaseMetadataIdentifier
 }
 
-// getIdentifierKey...
-func (mdi *MetadataIdentifier) getIdentifierKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.application)
+// getIdentifierKey will return string format as service:Version:Group:Side:Application
```
Comment (AlexStocks) [inline metadata/identifier/metadata_identifier.go:26]: getIdentifierKey? GetIdentifierKey.

File: metadata/identifier/metadata_identifier.go [hunk 28]
```diff
@@ -17,17 +17,18 @@
...
+func (mdi *MetadataIdentifier) GetIdentifierKey() string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.Application)
 }
 
-// getIdentifierKey...
-func (mdi *MetadataIdentifier) getFilePathKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getFilePathKey(mdi.application)
+// getFilePathKey will return string format as metadata/path/Version/Group/Side/Application
```
Comment (AlexStocks) [inline metadata/identifier/metadata_identifier.go:31]: getFilePathKey? GetFilePathKey.

File: metadata/identifier/service_metadata_identifier.go [hunk 29]
```diff
@@ -21,18 +21,19 @@ import (
...
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.protocol + constant.KEY_REVISON_PREFIX + mdi.revision)
+// getIdentifierKey will return string format as service:Version:Group:Side:Protocol:"revision"+Revision
+func (mdi *ServiceMetadataIdentifier) GetIdentifierKey() string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.Protocol, constant.KEY_REVISON_PREFIX+mdi.Revision)
 }
 
-// getIdentifierKey...
-func (mdi *ServiceMetadataIdentifier) getFilePathKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getFilePathKey(mdi.protocol + constant.KEY_REVISON_PREFIX + mdi.revision)
+// getFilePathKey will return string format as metadata/path/Version/Group/Side/Protocol/"revision"+Revision
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go:36]: getFilePathKey? GetFilePathKey.

File: metadata/identifier/service_metadata_identifier.go [hunk 30]
```diff
@@ -21,18 +21,19 @@ import (
...
-	protocol string
+	Revision string
+	Protocol string
 	BaseMetadataIdentifier
 }
 
-// getIdentifierKey...
-func (mdi *ServiceMetadataIdentifier) getIdentifierKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.protocol + constant.KEY_REVISON_PREFIX + mdi.revision)
+// getIdentifierKey will return string format as service:Version:Group:Side:Protocol:"revision"+Revision
```
Comment (AlexStocks) [inline metadata/identifier/service_metadata_identifier.go:31]: getIdentifierKey? GetIdentifierKey.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 31]
```diff
@@ -17,17 +17,18 @@
...
 type SubscriberMetadataIdentifier struct {
-	revision string
+	Revision string
 	BaseMetadataIdentifier
 }
 
-// getIdentifierKey...
-func (mdi *SubscriberMetadataIdentifier) getIdentifierKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.revision)
+// getIdentifierKey will return string format as service:Version:Group:Side:Revision
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go:26]: getIdentifierKey? getIdentifierKey.

File: metadata/identifier/subscribe_metadata_identifier.go [hunk 32]
```diff
@@ -17,17 +17,18 @@
...
+func (mdi *SubscriberMetadataIdentifier) GetIdentifierKey() string {
+	return mdi.BaseMetadataIdentifier.getIdentifierKey(mdi.Revision)
 }
 
-// getIdentifierKey...
-func (mdi *SubscriberMetadataIdentifier) getFilePathKey(params ...string) string {
-	return mdi.BaseMetadataIdentifier.getFilePathKey(mdi.revision)
+// getFilePathKey will return string format as metadata/path/Version/Group/Side/Revision
```
Comment (AlexStocks) [inline metadata/identifier/subscribe_metadata_identifier.go:31]: getIdentifierKey? getIdentifierKey.

File: metadata/namemapping/service_name_mapping.go (hunk 4 above)
Comment (AlexStocks) [reply metadata/namemapping/service_name_mapping.go:18]: mapping may be enough.

File: metadata/service/inmemory/service_test.go [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+	serviceName := "com.ikurento.user.UserProvider"
+	group := "group1"
+	version := "0.0.1"
+	protocol := "dubbo"
+	beanName := "UserProvider"
+	u, _ := common.NewURL(fmt.Sprintf("%v://127.0.0.1:20000/com.ikurento.user.UserProvider?anyhost=true&"+
+		"application=BDTService&category=providers&default.timeout=10000&dubbo=dubbo-provider-golang-1.0.0&"+
+		"environment=dev&interface=%v&ip=192.168.56.1&methods=GetUser&"+
+		"module=dubbogo+user-info+server&org=ikurento.com&owner=ZX&pid=1447&revision=0.0.1&"+
+		"side=provider&timeout=3000&timestamp=1556509797245&group=%v&version=%v&bean.name=%v", protocol, serviceName, group, version, beanName))
```
[the remark this answers is not in the document]
Comment (zouyx) [reply metadata/service/inmemory/service_test.go:70]: i think should use \`\`

File: metadata/service/inmemory/service.go [hunk 34]
```diff
@@ -0,0 +1,227 @@
...
+
+// GetSubscribedURLs get all subscribedUrl
+func (mts *MetadataService) GetSubscribedURLs() (sets.Set, error) {
+	return mts.getAllService(mts.subscribedServiceURLs), nil
+}
+
+// GetServiceDefinition can get service definition by interfaceName, group and version
+func (mts *MetadataService) GetServiceDefinition(interfaceName string, group string, version string) (string, error) {
+	serviceKey := definition.ServiceDescriperBuild(interfaceName, group, version)
+	v, _ := mts.serviceDefinitions.Load(serviceKey)
```
[the remark this answers is not in the document]
Comment (hxmhlt) [reply metadata/service/inmemory/service.go:209]: service key can not be nil.

File: metadata/service/inmemory/service.go [hunk 35]
```diff
@@ -0,0 +1,227 @@
...
+	return v.(string), nil
+}
+
+// GetServiceDefinition can get service definition by serviceKey
+func (mts *MetadataService) GetServiceDefinitionByServiceKey(serviceKey string) (string, error) {
+	v, _ := mts.serviceDefinitions.Load(serviceKey)
```
[the remark this answers is not in the document]
Comment (hxmhlt) [reply metadata/service/inmemory/service.go:215]: as above

File: common/rpc_service.go (hunk 25 above)
Comment (hxmhlt) [reply common/rpc_service.go:141]: @fangyincheng pls add it , thx.

Comment (hxmhlt) [reply common/rpc_service.go:141]: > pls add comment for this file. thx.
Notified Mr fang add it ,but not this pr.

File: config/config_loader_test.go [hunk 36]
```diff
@@ -18,6 +18,7 @@
 package config
 
 import (
+	"go.uber.org/atomic"
```
Comment (zouyx) [inline config/config_loader_test.go:21]: split

File: config/service_config.go [hunk 37]
```diff
@@ -160,22 +179,42 @@ func (c *ServiceConfig) Export() error {
...
+	}
+	if c.unexported.Load() {
+		return
+	}
+	c.exportersLock.Lock()
+	defer c.exportersLock.Unlock()
+	for _, exporter := range c.exporters {
+		exporter.Unexport()
+	}
+	c.exporters = nil
```
Comment (zouyx) [inline config/service_config.go:213]: just lock this block ,no need to  defer?

File: metadata/definition/definition.go [hunk 38]
```diff
@@ -39,3 +51,39 @@ type TypeDefinition struct {
...
+
+// ServiceDescriperBuild: build the service key, format is `group/serviceName:version` which be same as URL's service key
+func ServiceDescriperBuild(serviceName string, group string, version string) string {
+	buf := &bytes.Buffer{}
+	if group != "" {
+		buf.WriteString(group)
+		buf.WriteString(constant.PATH_SEPARATOR)
+	}
+	buf.WriteString(serviceName)
+	if version != "" && version != "0.0.0" {
```
Comment (zouyx) [inline metadata/definition/definition.go:84]: i dont suggest use "0.0.0" .

what about split by `.` and convert to int, finally compare to `0` ?

File: metadata/service/exporter/configurable/exporter.go [hunk 39]
```diff
@@ -0,0 +1,97 @@
...
+		exporter.serviceConfig = config.NewServiceConfig("MetadataService", context.Background())
+		exporter.serviceConfig.Protocol = constant.DEFAULT_PROTOCOL
+		exporter.serviceConfig.Protocols = map[string]*config.ProtocolConfig{
+			constant.DEFAULT_PROTOCOL: generateMetadataProtocol(),
+		}
+		exporter.serviceConfig.InterfaceName = constant.METADATA_SERVICE_NAME
+		exporter.serviceConfig.Group = config.GetApplicationConfig().Name
+		exporter.serviceConfig.Version = exporter.metadataService.Version()
+		exporter.serviceConfig.Implement(exporter.metadataService)
+		err := exporter.serviceConfig.Export()
```
Comment (zouyx) [inline metadata/service/exporter/configurable/exporter.go:62]:
```suggestion
		serviceConfig :=&config.ServiceConfig{}
		serviceConfig = config.NewServiceConfig("MetadataService", context.Background())
		serviceConfig.Protocol = constant.DEFAULT_PROTOCOL
		serviceConfig.Protocols = map[string]*config.ProtocolConfig{
			constant.DEFAULT_PROTOCOL: generateMetadataProtocol(),
		}
		serviceConfig.InterfaceName = constant.METADATA_SERVICE_NAME
		serviceConfig.Group = config.GetApplicationConfig().Name
		exporter.serviceConfig.Version = exporter.metadataService.Version()
		exporter.lock.Lock()
		exporter.serviceConfig=serviceConfig
		exporter.serviceConfig.Implement(exporter.metadataService)
		err := exporter.serviceConfig.Export()
		exporter.lock.Unlock()
```

File: metadata/service/inmemory/service.go [hunk 40]
```diff
@@ -0,0 +1,232 @@
...
+		defer mts.lock.RUnlock()
+		if value.(*skip.SkipList).Len() == 0 {
+			targetMap.Delete(url.ServiceKey())
+		}
+	}
+}
+
+// getAllService can return all the exportedUrlString except for metadataService
+func (mts *MetadataService) getAllService(services *sync.Map) *skip.SkipList {
+	skipList := skip.New(uint64(0))
```
Comment (zouyx) [inline metadata/service/inmemory/service.go:117]: as above

File: metadata/service/inmemory/service.go [hunk 41]
```diff
@@ -0,0 +1,232 @@
...
+			}
+		}
+		return true
+	})
+	return skipList
+}
+
+// getSpecifiedService can return specified service url by serviceKey
+func (mts *MetadataService) getSpecifiedService(services *sync.Map, serviceKey string, protocol string) *skip.SkipList {
+	skipList := skip.New(uint64(0))
```
Comment (zouyx) [inline metadata/service/inmemory/service.go:133]: as above

File: metadata/service/inmemory/service.go [hunk 42]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// addURL will add URL in memory
+func (mts *MetadataService) addURL(targetMap *sync.Map, url *common.URL) bool {
+	var (
+		urlSet interface{}
+		loaded bool
+	)
+	logger.Debug(url.ServiceKey())
+	if urlSet, loaded = targetMap.LoadOrStore(url.ServiceKey(), skip.New(uint64(0))); loaded {
```
Comment (zouyx) [inline metadata/service/inmemory/service.go:80]: why ```New(uint64(0))```? Should it change to a default value, like  `8` ?

File: metadata/service/service.go [hunk 43]
```diff
@@ -0,0 +1,63 @@
...
+	GetExportedURLs(serviceInterface string, group string, version string, protocol string) (*skip.SkipList, error)
+	// GetExportedURLs will get the target subscribed url in metadata
+	GetSubscribedURLs() (*skip.SkipList, error)
+	// GetServiceDefinition will get the target service info store in metadata
+	GetServiceDefinition(interfaceName string, group string, version string) (string, error)
+	// GetServiceDefinition will get the target service info store in metadata by service key
+	GetServiceDefinitionByServiceKey(serviceKey string) (string, error)
+	// Version will return the metadata service version
+	Version() string
+	common.RPCService
```
Comment (zouyx) [inline metadata/service/service.go:53]: what about put it to the front of all methods?

File: metadata/definition/definition.go (hunk 38 above)
Comment (hxmhlt) [reply metadata/definition/definition.go:84]: just be same as java

File: metadata/service/inmemory/service.go (hunk 40 above)
Comment (hxmhlt) [reply metadata/service/inmemory/service.go:117]: the value of param is not used. Just type is used.

**Pull request merged**